### PR TITLE
Specify shift modifier for default keybindings that implicitly require it

### PR DESCRIPTION
**Pull request by TAAPArthur**
This is a cosmetic change to add ShiftMask all naked capitol letters
keybindings. Don't think our defaults should hide the fact that the
shift modifier is used. Users can still not specify it if they prefer.

So now a binding like `{0, XK_A, ...}` would be defined like
`{ShiftMask, XK_A, ...}`, but something like `{0, XK_a, ...}` would be
unchanged.
<hr>
Original message:

Mainly to not break existing conventions, this commit requires ShiftMask
to be present in the keybindings when using second-level keysyms (like
capitol letters). This may break existing configs if users took
advantage or omitted ShiftModifier but all default bindings have been
updated in this commit.

So now a binding like `{0, XK_A, ...}` would be defined like
`{ShiftMask, XK_A, ...}`, but something like `{0, XK_a, ...}` would be
unchanged.

If the old behavior is desired of having the ShiftModifier be optional,
then simply adding ShiftMask to `ignore_mask` should suffice. See #77

### CODE REVIEW

**Comment by N-R-K** [comment]
Is there any practical benefit to this or is it just a style thing?
Because if it's the latter then I'm more in favor of, "If it aint broke dont fix it".

**Comment by TAAPArthur** [comment]
> Is there any practical benefit to this or is it just a style thing?

Style, readability, code quality etc, but no intrinsic benefit to the actual user. I view it more of code cleanup/standardization than a bug fix/new feature. Maybe I should have put it in its own commit, but there was a change to allow modifier keys like Shirt/Control to be used in keybindings which has practical use.

> Because if it's the latter then I'm more in favor of, "If it aint broke dont fix it".

Does that mean you want the same behavior or don't think it is worth touching the config file or just don't want the patch at all. I consider there to be 3 parts, 
1. the change to `on_keypress` which just makes it simpler.
2. The change in the default behavior to force ShiftMask to be specified.
3. Adding ShiftMask to all keybindings that "should" have it.

Think 1 is non-controversial and can be done without the others.

EDIT:
This is very opinionated but specifying ShiftMask makes it easier to see which modifiers to nsxiv uses by default so users can synchronize their keybindings for other applications (window managers, containers like tabbed, etc). When I glanced at the config I had thought on ControlMask and missed the implied ShiftMask. L

**Comment by N-R-K** [comment]
I'm not against the patch completely, just in favor of not changing things unless there's tangible benefit to be had. In this case, I'm not sure if there is any. But then again, we also changed the `enums` in config to `static const ints`.

I think I'm okay letting the other maintainers decide what to do here. Don't really have too strong of an opinion here. That's of course assuming this doesn't introduce some hidden/edge-case scenario where backwards compatibility gets broken.

**Comment by N-R-K** [comment]
On [surf](http://git.suckless.org/surf/file/config.def.h.html) and [tabbed](http://git.suckless.org/tabbed/file/config.def.h.html) there aren't any naked `ShiftMasks` like we have here, but where `ShiftMask` is used, it's explicit. I.e `Modkey|ShiftMaks`.

So I guess one can make an argument about this being more suckless. My opinion on this is still neutral, I'm just adding these info as they seem relevant. I will trust other maintainers' judgement on this and let them decide.

**Comment by bakkeby** [comment]
I agree. Explicitly stating the ShiftMask is more readable and intuitive (for me anyway).

Also that way you can combine it with the lower case keysym, e.g. ShiftMask, XK_a, ...

Whether that is more readable or not is subjective of course. I'm used to the lower case variant from suckless tools.

File: main.c
Comment (N-R-K) [inline main.c:589]: This was a bit silly, so removing this also has some readability benefit imo.

**Review by N-R-K — APPROVED** [review]
Given that it is more in line with other suckless tools, and it gets rid of a chunk of unneeded code, I think I've changed my mind on this. :+1:

**Comment by N-R-K** [comment]
@XPhyro https://github.com/nsxiv/nsxiv/pull/78/commits/153b0ca5633148f8377e100eaccbb759cbef2d7e breaks all the shift keybindings.

**Review by N-R-K — CHANGES_REQUESTED** [review]
ShiftMask keybindings are broken

**Comment by N-R-K** [comment]
This PR also breaks `key-handler`, since it how explicitly sends `Shift`.

**Comment by bakkeby** [comment]
So the `A` --> `a` change in the keybindings did not work because of this line?

```c
XLookupString(kev, &key, 1, &ksym, NULL);
```

In dwm they use XGetKeyboardMapping to get the symbolic key.

```c
keysym = XGetKeyboardMapping(dpy, (KeyCode)ev->keycode, 1, &keysyms_return);
```

@N-R-K how does it break the `key-handler`?

Based on the comment in the shell script it sounds more like this is fixing it. It would never get the `S-` prefix before.

```sh
# The key combo argument has the following form: "[C-][M-][S-]KEY",
# where C/M/S indicate Ctrl/Meta(Alt)/Shift modifier states and KEY is the X
# keysym as listed in /usr/include/X11/keysymdef.h without the "XK_" prefix.
```

**Comment by N-R-K** [comment]
@bakkeby currently on sxiv, if you press `shift+n` the key-handler would receive `N`. If we want to keep it that way then the expected behaviour should be `N`, otherwise if we want to follow the comment instead then the expected behaviour should be `S-n`

However currently this PR does neither. It simply sends `ShiftL` or `ShiftR`.

**Comment by bakkeby** [comment]
>  It simply sends ShiftL or ShiftR. The key after shift doesn't get sent to the key-handler.

I see what you mean, the modifier keypress is sent as a key to the key-handler script. This happens for Alt and Ctrl as well. It seems to only work if you keep holding down Ctrl after having done Ctrl+x to trigger the external script. 

This seems to work for me (lower case syms + modifiers being passed to the key-handler script correctly), but I don't like the ugly block checking if the key pressed is itself a modifier key. There must be a better way to determine that.

```c
void on_keypress(XKeyEvent *kev)
{
	int i;
	KeySym *ksym;
	int keysyms_return;
	char *key;
	bool dirty = false;

	ksym = XGetKeyboardMapping(win.env.dpy, (KeyCode)kev->keycode, 1, &keysyms_return);
	key = XKeysymToString(*ksym);

	if (*ksym == XK_Shift_L ||
	    *ksym == XK_Shift_R ||
	    *ksym == XK_Control_L ||
	    *ksym == XK_Control_R ||
	    *ksym == XK_Alt_L ||
	    *ksym == XK_Alt_R
	) {
		XFree(ksym);
		return;
	}

	if (*ksym == XK_Escape && MODMASK(kev->state) == 0) {
		extprefix = False;
	} else if (extprefix) {
		run_key_handler(key, kev->state);
		extprefix = False;
	} else if (*key >= '0' && *key <= '9') {
		/* number prefix for commands */
		prefix = prefix * 10 + (int) (*key - '0');
		return;
	} else for (i = 0; i < ARRLEN(keys); i++) {
		if (keys[i].ksym == *ksym &&
		    MODMASK(keys[i].mask) == MODMASK(kev->state) &&
		    keys[i].cmd >= 0 && keys[i].cmd < CMD_COUNT &&
		    (cmds[keys[i].cmd].mode < 0 || cmds[keys[i].cmd].mode == mode))
		{
			if (cmds[keys[i].cmd].func(keys[i].arg))
				dirty = true;
		}
	}
	XFree(ksym);
	if (dirty)
		redraw();
	prefix = 0;
}
```

**Comment by N-R-K** [comment]
> > It simply sends ShiftL or ShiftR. The key after shift doesn't get sent to the key-handler.
> 
> I see what you mean, the modifier keypress is sent as a key to the key-handler script. This happens for Alt and Ctrl as well. It seems to only work if you keep holding down Ctrl after having done Ctrl+x to trigger the external script.
> 
> This seems to work for me (lower case syms + modifiers being passed to the key-handler script correctly), but I don't like the ugly block checking if the key pressed is itself a modifier key. There must be a better way to determine that.
> 
> ```c
> void on_keypress(XKeyEvent *kev)
> {
> 	int i;
> 	KeySym *ksym;
> 	int keysyms_return;
> 	char *key;
> 	bool dirty = false;
> 
> 	ksym = XGetKeyboardMapping(win.env.dpy, (KeyCode)kev->keycode, 1, &keysyms_return);
> 	key = XKeysymToString(*ksym);
> 
> 	if (*ksym == XK_Shift_L ||
> 	    *ksym == XK_Shift_R ||
> 	    *ksym == XK_Control_L ||
> 	    *ksym == XK_Control_R ||
> 	    *ksym == XK_Alt_L ||
> 	    *ksym == XK_Alt_R
> 	) {
> 		XFree(ksym);
> 		return;
> 	}
> 
> 	if (*ksym == XK_Escape && MODMASK(kev->state) == 0) {
> 		extprefix = False;
> 	} else if (extprefix) {
> 		run_key_handler(key, kev->state);
> 		extprefix = False;
> 	} else if (*key >= '0' && *key <= '9') {
> 		/* number prefix for commands */
> 		prefix = prefix * 10 + (int) (*key - '0');
> 		return;
> 	} else for (i = 0; i < ARRLEN(keys); i++) {
> 		if (keys[i].ksym == *ksym &&
> 		    MODMASK(keys[i].mask) == MODMASK(kev->state) &&
> 		    keys[i].cmd >= 0 && keys[i].cmd < CMD_COUNT &&
> 		    (cmds[keys[i].cmd].mode < 0 || cmds[keys[i].cmd].mode == mode))
> 		{
> 			if (cmds[keys[i].cmd].func(keys[i].arg))
> 				dirty = true;
> 		}
> 	}
> 	XFree(ksym);
> 	if (dirty)
> 		redraw();
> 	prefix = 0;
> }
> ```

Even with this `shift+=` aka `+` is broke, `{` and `}` are also broken.

**Comment by bakkeby** [comment]
Yes you are probably right in that it might be more trouble than it is worth.

```c
void on_keypress(XKeyEvent *kev)
{
	int i;
	KeySym *ksym, extksym;
	int keysyms_return;
	char key;
	bool dirty = false;

	ksym = XGetKeyboardMapping(win.env.dpy, (KeyCode)kev->keycode, 1, &keysyms_return);
	XLookupString(kev, &key, 1, &extksym, NULL);

	if (IsModifierKey(extksym)) {
		XFree(ksym);
		return;
	}

	if (*ksym == XK_Escape && MODMASK(kev->state) == 0) {
		extprefix = False;
	} else if (extprefix) {
		run_key_handler(XKeysymToString(extksym), kev->state);
		extprefix = False;
	} else if (key >= '0' && key <= '9') {
		/* number prefix for commands */
		prefix = prefix * 10 + (int) (key - '0');
		return;
	} else for (i = 0; i < ARRLEN(keys); i++) {
		if (keys[i].ksym == *ksym &&
		    MODMASK(keys[i].mask) == MODMASK(kev->state) &&
		    keys[i].cmd >= 0 && keys[i].cmd < CMD_COUNT &&
		    (cmds[keys[i].cmd].mode < 0 || cmds[keys[i].cmd].mode == mode))
		{
			if (cmds[keys[i].cmd].func(keys[i].arg))
				dirty = true;
		}
	}
	XFree(ksym);
	if (dirty)
		redraw();
	prefix = 0;
}
```

With the above the external keys work as they do in sxiv, but then we are using both `XGetKeyboardMapping` and `XLookupString` to get the symbolic key. So in the key-handler script you'd get `S-braceright` but to have the same key combination as an internal keybinding you would have to use `ShiftMask, XK_bracketleft` (rather than `XK_braceleft`). This is just going to be a source of confusion I feel.

So I think we should just leave the `A` --> `a` complication out of it and just add back in lines 588-589 that was removed.

```c
	if (IsModifierKey(ksym))
		return;
```

With that I think the PR works as intended also for external scripts.

**Comment by N-R-K** [comment]
> With that I think the PR works as intended also for external scripts.

Almost, it sends `S-R` , if we're going to be explicitly doing `S-`, then I feel like it should be `S-r` instead.
Same on `config.def.h`, if we do explicit `ShiftMask` then the keys should be expected to be lower case...

Initially I wasn't against this after seeing that other suckless projects are using explicit `ShiftMask` but now I've sorta changed my mind once again. Imo we shouldn't break user's existing `key-handler` for what is basically a cosmetic/style change for the most part.

**Comment by XPhyro** [comment]
> While I do agree that explicit `ShiftMask` is better, this may end up being more trouble than it's worth.

This pretty much sums up my thoughts about this PR.

- While I do like the explicit ShiftMask, I do not like having to also write the key as shifted. The latter restriction completely eliminates the point of ShiftMask and creates confusion for those already used to suckless software.
- With or without that given snippet, we break backwards compatibility with the already existing configs and key handlers. I am content with breaking the existing configs given the benefit of ShiftMask (if the restriction is resolved), but this should not break key handlers. It is more convenient and conventional to have uppercase letters instead of the S prefix. And, if anything, it should be S-lowercase instead of S-Uppercase.

**Comment by bakkeby** [comment]
Perhaps we should review the whole external key handler thing, it seems like more of a hack. There is also https://github.com/muennich/sxiv/pull/428 that allows multiple commands to go to the external key handler without having to do Ctrl+x each time.

**Comment by N-R-K** [comment]
> Perhaps we should review the whole external key handler thing, it seems like more of a hack.

I'm open for that discussion, how should we go about it then? However if we are to break backwards compatibility (which is against the project scope) there should be *tangible* benefit to be had, not just style change.

> allows multiple commands to go to the external key handler without having to do Ctrl+x each time.

Not a fan of this, but don't totally oppose it either. Seems more like a personal/niche thing.

**Comment by XPhyro** [comment]
I don't think there is any benefit to be gained by reworking key-handler. The reworking of the S prefix is just a style change, breaks backwards compatibility and breaks the general notion in terminal applications to use the uppercase equivalent instead of the S prefix.

I am okay with the `multiple commands` thing, as that can be implemented in the nsxiv source without breaking existing key handlers.

For the "unnecessary condition" on the Escape key, see #91.

**Comment by bakkeby** [comment]
On my keyboard layout a capital `A` is typed by `Shift+a` (not considering Caps Lock here).

Likewise a braceleft `{` is typed by holding the Shift modifier and pressing the `]` (bracketleft) key.

The code as-is hides that the Shift modifier is needed to get to a key like capital `A` or a braceleft as in this example.

If we make it so that the ShiftMask must be explicitly set for second level symbolic keys then that means that we end up with external key strings in the key-handler such as `S-A` and `S-braceleft`.

If it makes more sense that the actual first level symbolic keys are presented to the external key-handler then we will end up with `S-a` and `S-bracketleft`. Expecting a mix of the two where plain letters are lower case while other characters like the braceleft remain on second level doesn't sound very logical to me.

So I think what this PR and discussion boils down to is to simply remove support for second level keysyms in keybindings (making it in line with other suckless tools that only supports first level keysyms). This is a change that I personally would be in favour for.

That said given the concerns about backwards compatibility we should probably just close / reject this.

I think it is worth noting the other side of this which is that because of how the existing code works there is no support for first level keysyms when a second level keysym exists.

**Comment by TAAPArthur** [comment]
> That said given the concerns about backwards compatibility we should probably just close / reject this.

If backwards compatibility is the main issue, then one could just add `ShiftMask` to `ignore_mask`. It will give the same behavior as before. We already break backwards compatibility with any new option so there being an easy way to get the old behavior would minimize that concern in my opinion.

Also maybe I should have made it a separate commit but
```
if (IsModifierKey(ksym))
	return;
```
was removed which is what is causing modifier keys to be sent to key handler/keybindings. It is tangential to the main change (forcing ShiftMask to be specified) and can be its own PR/discussion if it is controversial. Binding to the modifier key can be useful if you want to do something for all your bindings involving that modifier.

> Perhaps we should review the whole external key handler thing, it seems like more of a hack.

I believe that https://github.com/nsxiv/nsxiv/pull/76 is strictly more versatile and easier to use than the current key handlers.

**Comment by N-R-K** [comment]
> ```
> if (IsModifierKey(ksym))
> 	return;
> ```
> 
> was removed which is what is causing modifier keys to be sent to key handler/keybindings.

Adding that would still send `S-R` to the keyhandler, breaking backwards-compatibility. Also as me and Xphyro already mentioned, sending explicit `S-` but then ALSO sending capital letter is a bit illogical.

- - -

If there aren't any objections, then we can close this PR. See [this](https://github.com/nsxiv/nsxiv/pull/78#issuecomment-925631653) comment for rational.

**Comment by TAAPArthur** [comment]
>Adding that would still send S-R to the keyhandler, breaking backwards-compatibility. 

"If backwards compatibility is the main issue, then one could just add ShiftMask to ignore_mask." Should get the same behavior as before this change. Might need change the line to 
``` 		run_key_handler(XKeysymToString(ksym), kev->state & ~ignore_mask);```
If we think this is a better format going forward, we could proceed with this change and ShiftMask to ignore_mask to keep previous behavior. Or just tell people to add it if they are actually affected by it.

> Also as me and Xphyro already mentioned, sending explicit `S-` but then ALSO sending capital letter is a bit illogical.

You say illogical, but that's a more accurate representation of what has happened. We do extra work to try to change this "natural" representation. And we only do it for letters. When "!" is sent you don't receive a "1" with a ShiftModifier. It is consistent with all the other non-alphabetic keys so don't think labeling is "illogical" is fair. I had thought the main object was backwards compatibility rather than the change in isolation

**Comment by N-R-K** [comment]
> When "!" is sent you don't receive a "1" with a ShiftModifier.

But you're pressing "1" with a ShiftModifier. Same with "A", its `shift+a = A`, *not* `shift+A = A`.

Also since this is mostly just a style change, is there really much reason for pushing for this?

**Comment by TAAPArthur** [comment]
> Also since this is mostly just a style change, is there really much reason for pushing for this?

No. Allowing modifier keys to be bound may have some use in theory. Everything else is just readability. I just don't think backwards compatibility (in the feature sense) should be the reason to not do this since we could work around that easily. If we just don't want to cosmetic refactoring for the sake or readability or the claim of readability was in question or something else to a similar effect, think it would be fine to close. 

What is intuitive about the behavior may be subjective. I can see your view as looking at it as pressing "a" with "shift" held. I think of it as receiving "A" with "shift" held. This discussion has given me a better appreciation for programs that just use the translate the key sym to a key code (so using "A" or "a" is exactly the same) and this avoid this paradigm difference

**Comment by TAAPArthur** [comment]
So relooking at the diff. Aside from the config file, think `on_keypress` looks cleaner with this.

@N-R-K Do you expect "!" to be sent with or without a shift? If you expect it with a shift but still want "A" without then the solution of changing ignore_mask wouldn't be sufficient.

**Pull request closed without merging**
